cloudflare/resource_cloudflare_waf_package.go

```diff
@@ -0,0 +1,255 @@
+			return err
+		}
+	}
+
+	// We need to perform the validation inside the `Create` to ensure
+	// we get at least one of the expected values.
+	if packageName == "" && packageID == "" {
+		return fmt.Errorf("either package name or ID must be provided")
+	}
+
+	if packageID == "" {
+		pkg, err := packageFromName(client, zoneID, packageName)
+		if err != nil {
+			return err
+		}
+		packageID = pkg.ID
+	}
+
+	pkg, err := client.WAFPackage(zoneID, packageID)
+	if err != nil {
+		return fmt.Errorf("Unable to find WAF Package %s", packageID)
+	}
+
+	d.Set("zone_id", zoneID)
+	d.Set("package_id", packageID)
+	d.Set("package_name", pkg.Name)
+	d.Set("sensitivity", sensitivity)
+	d.Set("action_mode", actionMode)
+
+	// Set the ID to the package_id parameter passed in from the user.
+	// All WAF packages already exist so we already know the package_id.
+	//
+	// This is a work around as we are not really "creating" a WAF Package,
+	// only associating it with our terraform config for future updates.
+	d.SetId(packageID)
+
+	if pkg.Sensitivity != sensitivity || pkg.ActionMode != actionMode {
+		err = resourceCloudflareWAFPackageUpdate(d, meta)
+		if err != nil {
+			d.SetId("")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func resourceCloudflareWAFPackageDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+
+	packageID := d.Get("package_id").(string)
+	zoneID := d.Get("zone_id").(string)
+
+	pkg, err := client.WAFPackage(zoneID, packageID)
+	if err != nil {
+		return err
+	}
+
+	// Can't delete WAF Package so instead reset it to default
+	schema := resourceCloudflareWAFPackage().Schema
+	defaultSensitivity := schema["sensitivity"].Default.(string)
+	defaultActionMode := schema["action_mode"].Default.(string)
+
+	if pkg.Sensitivity != defaultSensitivity || pkg.ActionMode != defaultActionMode {
+		options := cloudflare.WAFPackageOptions{
+			Sensitivity: defaultSensitivity,
+			ActionMode:  defaultActionMode,
+		}
+
+		_, err = client.UpdateWAFPackage(zoneID, packageID, options)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func resourceCloudflareWAFPackageUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+
+	packageID := d.Get("package_id").(string)
+	zoneID := d.Get("zone_id").(string)
+	sensitivity := d.Get("sensitivity").(string)
+	actionMode := d.Get("action_mode").(string)
+
+	options := cloudflare.WAFPackageOptions{
+		Sensitivity: sensitivity,
+		ActionMode:  actionMode,
+	}
+
+	_, err := client.UpdateWAFPackage(zoneID, packageID, options)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceCloudflareWAFPackageImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	client := meta.(*cloudflare.API)
+
+	// split the id so we can lookup
+	idAttr := strings.SplitN(d.Id(), "/", 2)
+	var zoneName string
+	var packageNameOrID string
+	if len(idAttr) == 2 {
+		zoneName = idAttr[0]
+		packageNameOrID = idAttr[1]
```

Comment:
What cases isn't a package ID available? Ideally we don't use names for references in case they ever change inside of Cloudflare and that doesn't reflect in the terraform state.

Reply:
Packages ID do not seem easily available through the interface (same for groups in the other PR), which is why I allowed matching through the name.
WAF rules IDs are easily accessible (you can see them right when clicking on a WAF Group), but WAF Groups IDs and WAF Packages IDs are not?

Reply:
I don't think it's really viable. The name has spaces and is subject to change if the package ends up renamed. The IDs can be pulled from the API so perhaps we just include the documentation for fetching it from there for importing? The [`cloudflare/cf-terraforming`](https://github.com/cloudflare/cf-terraforming/) project might add support for this and look after those calls for the user too.

Reply:
As you wish, I can change it.
This will make the terraform configuration less easy to read though, but if there's tooling around it, why not